### PR TITLE
Add http2 options and updated tyk conf

### PR DESCRIPTION
**Pull request by marksou**
Updated gRPC intro page and tyk.conf settings. Need any Environment Variables to add to the Gateway Env Variables page.

### CODE REVIEW

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md [hunk 1]
```diff
@@ -235,6 +235,7 @@ Set these options to hard-code values into the way the HTTP server behaves.
 
 ```
 "http_server_options": {
+  "enable_http2": false
```
Comment (furkansenharputlu) [inline tyk-docs/content/configure/tyk-gateway-configuration-options.md]: This value should be `true` by default. Also, there is a missing `,` at the end.

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md [hunk 2]
```diff
@@ -457,18 +463,22 @@ For additional security it is possible to have Tyk put its REST API on a separat
...
 
 ### <a name="enable_custom_domains"></a> enable_custom_domains
 
 Set this value to `true` to enable this node to bind APIs to custom domains set in the API definition.
 
-### <a name="hostname"></a> hostname
+### <a name="proxy_enable_http2"></a> proxy_enable_http2
 
-The hostname to bind the node to. If set, all API traffic must go via this host name, otherwise it will raise a 404.
+Set this to `true` for HTTP 2 upstream connections.
```
Comment (furkansenharputlu) [inline tyk-docs/content/configure/tyk-gateway-configuration-options.md]: Same for this. Its default value should be `true`. Also, `HTTP/2`.

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md [hunk 3]
```diff
@@ -254,6 +255,11 @@ Set these options to hard-code values into the way the HTTP server behaves.
   "ssl_insecure_skip_verify": false
 },
 ```
+
+#### <a name="enable_http2"></a> enable_http2
+
+Set to true for HTTP 2 connections.
```
Comment (furkansenharputlu) [inline tyk-docs/content/configure/tyk-gateway-configuration-options.md]: The general usage is `HTTP/2`.

File: tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md [hunk 4]
```diff
@@ -7,11 +7,30 @@ menu:
...
 
+### How can I use gRPC with Tyk?
+
+Using Tyk with your gRPC client and server is very easy. Since gRPC uses HTTP/2, you need to enable it by setting `enable_http2=true` for `Downstream-Tyk` and `proxy_enable_http2=true` for `Tyk-Upstream` connections. You also need to set your `listen_path` in your API definitions. 
+
+### Mutual Authentication
+Tyk supports Mutual Authentication in gRPC. See https://tyk.io/docs/security/tls-and-ssl/mutual-tls/ to configure Mutual Authentication in Tyk. Now, everything will be same as default Mutual Authentication.
+
+### Basic Authentication
+Tyk supports Basic Authentication in gRPC. See https://tyk.io/docs/security/your-apis/basic-auth/ to configure Basic Authentication in Tyk.
```
Comment (furkansenharputlu) [inline tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md]: Maybe we can hide links like this:
See [Basic Authentication](https://tyk.io/docs/security/your-apis/basic-auth/) section to configure ...

Comment (marksou) [reply tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md]: Changed in new commit

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md (hunk 3 above)
Comment (marksou) [reply tyk-docs/content/configure/tyk-gateway-configuration-options.md]: Changed in new commit

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md (hunk 2 above)
Comment (marksou) [reply tyk-docs/content/configure/tyk-gateway-configuration-options.md]: Changed in a new commit.

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md (hunk 1 above)
Comment (marksou) [reply tyk-docs/content/configure/tyk-gateway-configuration-options.md]: Changed in a new commit.

**Review by furkansenharputlu — APPROVED** [review]

**Comment by furkansenharputlu** [comment]
@buger I think new env vars will be:
`TYK_GW_HTTPSERVEROPTIONS_ENABLEHTTP2`
`TYK_GW_PROXYENABLEHTTP2` 
right?

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md [hunk 5]
```diff
@@ -457,18 +463,22 @@ For additional security it is possible to have Tyk put its REST API on a separat
...
-### <a name="hostname"></a> hostname
+### <a name="proxy_enable_http2"></a> proxy_enable_http2
 
-The hostname to bind the node to. If set, all API traffic must go via this host name, otherwise it will raise a 404.
+This defaults to `true` for HTTP/2 upstream connections.
+
+
+### <a name="enable_jsvm"></a> enable_jsvm
+
+By default we have now disabled the JavaScript middleware system to ensure higher performance on nodes. If you are using the JSVM (custom middleware, or virtual endpoints), then enable this setting.
```
Comment (asoorm) [inline tyk-docs/content/configure/tyk-gateway-configuration-options.md:481]: > By default we have now disabled the JavaScript middleware system to ensure higher performance on nodes.

Does this not belong inside the changelog?

Could be reworded to:

`enable_jsvm` is false by default. Enable if you are using JSVM custom middleware or virtual endpoints.

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md [hunk 6]
```diff
@@ -254,6 +255,11 @@ Set these options to hard-code values into the way the HTTP server behaves.
   "ssl_insecure_skip_verify": false
 },
 ```
+
+#### <a name="enable_http2"></a> enable_http2
+
+This defaults to true for HTTP/2 connections.
```
Comment (asoorm) [inline tyk-docs/content/configure/tyk-gateway-configuration-options.md:261]: I don't understand what this means. config option `enable_http2` implies that I have to explicitly enable http2.

Yet the description `This defaults to true for HTTP/2 connections.` implies that true is the default. If `enable_http2: true` is default, then maybe config option should be `disable_http2`?

File: tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md [hunk 7]
```diff
@@ -7,11 +7,30 @@ menu:
...
 ---
-### What is gRPC?
+## What is gRPC?
 From the [gRPC documentation][1]:
 
 > gRPC is a modern, open source remote procedure call (RPC) framework that can run anywhere. It enables client and server applications to communicate transparently, and makes it easier to build connected systems.
 
+### How can I use gRPC with Tyk?
+
+Using Tyk with your gRPC client and server is very easy. Since gRPC uses HTTP/2, you need to enable it by setting `enable_http2=true` for `Downstream-Tyk` and `proxy_enable_http2=true` for `Tyk-Upstream` connections. You also need to set your `listen_path` in your API definitions.
```
Comment (asoorm) [inline tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md:17]: Are we able to use http2/gRPC insecure skip verify or without https? Is it just http or can the gRPC upstream server be tcp/unix socket? What does the connection string look like?

Comment (asoorm) [reply tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md:17]: Also, terminology for `Downstream-Tyk` and `Tyk-Upstream` is a bit confusing.

File: tyk-docs/content/configure/tyk-gateway-configuration-options.md (hunk 6 above)
Comment (buger) [reply tyk-docs/content/configure/tyk-gateway-configuration-options.md:261]: Our default configs have this option enabled. So all new installations have it, and old will have to explicitly enable it.

File: tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md (hunk 7 above)
Comment (buger) [reply tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md:17]: Only http2 protocol supported, and http2 works only with TLS. And our standard TLS options will work.

Comment (buger) [reply tyk-docs/content/customise-tyk/plugins/rich-plugins/grpc/grpc.md:17]: Agree about confusion. Can be re-worded to smth like this:

Tyk supports gRPC pass hough proxying when HTTP/2 used as a transport (the most common way to deploy gRPC services). So the only requirement is enabling HTTP/2 support on the gateway side, for both incoming and upstream connections, by setting `http_server_options.enable_http2` and `proxy_enable_http2` to `true`.

[gRPC over HTTP2 specification](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) define rules on how gRPC protocol maps to HTTP request. In context of the API Gateway, we are intersted in the following things:
- HTTP path follow the format `/{Service-Name}/{method name}`, for example: `/google.pubsub.v2.PublisherService/CreateTopic`. You can use this feature to apply standard ACL rules via Keys and Policies, or use url rewrite plugins in endpoint desiger.
- HTTP method always `POST`
- gRPC custom request metadata added as HTTP headers, where metadata key directly mapped to HTTP header with the same name.

**Comment by buger** [comment]
Whole updates from this PR inside plugins/rich-plugins/grpc/grpc.md should be moved to a separate page, because gRPC proxying do not have any relation to our gRPC plugins. This is a totally separate feature.

**Comment by buger** [comment]
Needs additional paragraph on that Tyk can be used as a facade level, e.g. url rewriter can be used. So different services urls can point to different grpc servers. 

Also add code example for go-grpc, for basic auth part, showing that it just sets metadata.

**Pull request merged**